src/net_processing.cpp

```diff
@@ -1563,6 +1563,11 @@ void static ProcessGetData(CNode* pfrom, const CChainParams& chainparams, CConnm
             if (mi != mapRelay.end()) {
                 connman->PushMessage(pfrom, msgMaker.Make(nSendFlags, NetMsgType::TX, *mi->second));
                 push = true;
+                // Once the first peer requests GETDATA for a txn, we deem initial broadcast a success
```

Comment:
Not sure if one `GETDATA` is the best solution to indicate the txn has been successfully broadcast to the network. 

The worst case is highlighted by the tests- a node receives the getdata back from the first peer before announcing to any other peers. so I have to disconnect the other peers (see [here](https://github.com/bitcoin/bitcoin/pull/16698/files?file-filters%5B%5D=.py&file-filters%5B%5D=.sh#diff-e4786db1a6d286db4b104018b9b77608R70))

in practice, this means a user who submits a txn with p2p disabled, then re-enables, might only get the txn out to _one_ other peer before marking it as "broadcast"

Reply:
some relevant info on how an attack could potentially be carried out here: https://github.com/bitcoin/bitcoin/pull/16698#discussion_r350450955. but relevant parts reiterated here- 

the attack seems simple to execute, but rare to get the right conditions where it succeeds, and the incentive seems weak. 
- execution: attacker sends `GETDATAs` back for every transaction it receives & never propagate them out to the network. 
- conditions needed: 
1. node would be relying on unbroadcast set to broadcast txn to network (attacker couldn't discern) 
2. node would have to relay to attacker first 
3. timing would have to be fast enough that node receives & processes `GETDATA` before sending out other `INVs`. 
- outcome: the transaction doesn't propagate until the rebroadcast logic picks it up & sends it out to all peers. the attacker delayed the initial propagation of the transaction. 

doesn't seem like a very worthwhile attack to me.... 
seems much more likely to occur under accidental conditions. 

but I was thinking it through & am sharing incase reviewers want to weigh in.

Reply:
I find the mechanism a bit weird as it leaks a wallet concern in the P2P stack. If you have multiple wallets connected to your node or even accept tx from third-party through your RPC command are you going to ensure success for all of them ?

Reply:
I agree that its not an ideal separation of concerns, but I think it's reasonable. Essentially, when you submit a txn to the mempool/node, it guarantees that it will 1. validate the txn & accept into its mempool locally & 2. propagate it out to the network (assuming validity). 

The normal way this is achieved is by bundling the actions together. However, if you have p2p disabled when you initially submit the txn, you no longer get #2. 

Previously, this wasn't an issue because the rebroadcast logic was so over-eager, that initial relay would automatically occur within a short amount of time after enabling p2p. With these changes, that is no longer a guarantee, so we need to explicitly keep track of the transactions that have passed step #1, but not yet #2. 

The other option would be that the mempool returns indication when the txn was accepted locally & then when the txn was propagated to the network. Then the entity submitting the txn (wallet, third-party RPC, whatever) would have to keep track of txns in the in-between state. I think thats a much messier interaction between the layers.

Reply:
> 1. validate the txn & accept into its mempool locally 

Assuming tx succeed fee checks.

> The normal way this is achieved is by bundling the actions together. However, if you have > p2p disabled when you initially submit the txn, you no longer get # 2.

As a user if you deactivate p2p functionality at tx mempool-insertion broadcast should be then a best-effort on the timing side. I think this point is really linked with wallet timer rebroadcast, if you reduce it to be under the mempool one, won't you be sure than your txn is going to be initially broadcast at least within 1 hour ?

> The other option would be that the mempool returns indication when the txn was accepted locally & then when the txn was propagated to the network. Then the entity submitting the txn (wallet, third-party RPC, whatever) would have to keep track of txns in the in-between state. I think thats a much messier interaction between the layers.

If a wallet gets notification of transaction mempool insertion it should rely on it for network propagation but mempool may fail for its own internal reasons (size, conflicts) so for reliability your wallet should ensure periodic resend/rebroadcast.

Reply:
@ariard do you still have an outstanding concern here?

>I find the mechanism a bit weird as it leaks a wallet concern in the P2P stack.
the first comment talks about the layer separation
>so for reliability your wallet should ensure periodic resend/rebroadcast.
but your second comment seems to talk more about the wallet & mempool rebroadcast timers. 

the thread was started in regards to the *unbroadcast* logic. which is different than the *rebroadcast* logic. I explained the reason that it makes sense for the node to keep track of txns that are being initially relayed, let me know if you have further questions. I believe I have addressed your concerns around the timers elsewhere on this PR.

I'm going to resolve this conversation. If there's further questions or concerns about the unbroadcast logic I can re-open.